src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java

```diff
@@ -0,0 +1,298 @@
+        if (!isAbleToTriggerEviction &&
+                !cache.asMap().containsKey(nativeMemoryEntryContext.getKey()) &&
+                maxWeight - getCacheSizeInKilobytes() - nativeMemoryEntryContext.calculateSizeInKb() <= 0
+        ) {
+            throw new OutOfNativeMemoryException("Failed to load \"" + nativeMemoryEntryContext.getKey() +
+                    "\" into memory.");
+        }
+
+        return cache.get(nativeMemoryEntryContext.getKey(), nativeMemoryEntryContext::load);
+    }
+
+    /**
+     * Invalidate entry from the cache.
+     *
+     * @param key Identifier of entry to invalidate
+     */
+    public void invalidate(String key) {
+        cache.invalidate(key);
+    }
+
+    /**
+     * Invalidate all entries in the cache.
+     */
+    public void invalidateAll() {
+        cache.invalidateAll();
+    }
+
+    /**
+     * Returns whether or not the capacity of the cache has been reached
+     *
+     * @return Boolean of whether cache limit has been reached
+     */
+    public Boolean isCacheCapacityReached() {
+        return cacheCapacityReached.get();
+    }
+
+    /**
+     * Sets cache capacity reached
+     *
+     * @param value Boolean value to set cache Capacity Reached to
+     */
+    public void setCacheCapacityReached(Boolean value) {
+        cacheCapacityReached.set(value);
+    }
+
+    /**
+     * Get the stats of all of the Elasticsearch indices currently loaded into the cache
+     *
+     * @return Map containing all of the Elasticsearch indices in the cache and their stats
+     */
+    public Map<String, Map<String, Object>> getIndicesCacheStats() {
+        Map<String, Map<String, Object>> statValues = new HashMap<>();
+        NativeMemoryAllocation.IndexAllocation indexAllocation;
+
+        for (Map.Entry<String, NativeMemoryAllocation> entry : cache.asMap().entrySet()) {
+
+            if (entry.getValue() instanceof NativeMemoryAllocation.IndexAllocation) {
+                indexAllocation = (NativeMemoryAllocation.IndexAllocation) entry.getValue();
+                String indexName = indexAllocation.getOpenSearchIndexName();
+
+                Map<String, Object> indexMap = statValues.computeIfAbsent(indexName, name -> new HashMap<>());
+                indexMap.computeIfAbsent(GRAPH_COUNT, key -> getIndexGraphCount(indexName));
+                indexMap.computeIfAbsent(StatNames.GRAPH_MEMORY_USAGE.getName(), key ->
+                        getIndexSizeInKilobytes(indexName));
+                indexMap.computeIfAbsent(StatNames.GRAPH_MEMORY_USAGE_PERCENTAGE.getName(), key ->
+                        getIndexSizeAsPercentage(indexName));
+            }
+        }
+
+        return statValues;
+    }
+
+
+    private void onRemoval(RemovalNotification<String, NativeMemoryAllocation> removalNotification) {
+        NativeMemoryAllocation nativeMemoryAllocation = removalNotification.getValue();
+        executor.execute(() -> {
```

Comment:
I think its good idea to have 2 executors one for Training allocation and one for Indexing allocation as its possible that acquiring write lock in case of training could potentially block freeing up all other index allocations  and cause memory outage issues. Lets add two single threaded executors

Reply:
Sure, I can add this.